File: core/src/main/java/nl/t64/game/rpg/screens/inventory/CalcsTable.java
package nl.t64.game.rpg.screens.inventory;

import nl.t64.game.rpg.components.party.CalcType;
import nl.t64.game.rpg.components.party.HeroItem;
import nl.t64.game.rpg.components.party.InventoryGroup;


class CalcsTable extends BaseTable {

    private static final float FIRST_COLUMN_WIDTH = 190f;
    private static final float SECOND_COLUMN_WIDTH = 40f;
    private static final float THIRD_COLUMN_WIDTH = 35f;

    CalcsTable() {
        this.table.columnDefaults(0).width(FIRST_COLUMN_WIDTH);
        this.table.columnDefaults(1).width(SECOND_COLUMN_WIDTH);
        this.table.columnDefaults(2).width(THIRD_COLUMN_WIDTH);
        setTopBorder(this.table);
    }

    void update() {
        HeroItem selectedHero = InventoryUtils.selectedHero;
        table.clear();

        table.add("Weight");
        table.add("?");
        table.add("").row();

        table.add("Movepoints");
        table.add("?");
        table.add("").row();

        table.add(CalcType.BASE_HIT.getTitle());
        table.add(String.format("%s%%", selectedHero.getCalcValueOf(InventoryGroup.WEAPON, CalcType.BASE_HIT)));
        table.add("").row();

        table.add(CalcType.DAMAGE.getTitle());
        table.add(String.valueOf(selectedHero.getCalcValueOf(InventoryGroup.WEAPON, CalcType.DAMAGE)));
        table.add("").row();

        table.add(String.format("Total %s", CalcType.PROTECTION.getTitle()));
        table.add(String.valueOf(selectedHero.getTotalCalcOf(CalcType.PROTECTION)));
        table.add("").row();

        table.add(String.format("Shield %s", CalcType.PROTECTION.getTitle()));
        table.add(String.valueOf(selectedHero.getCalcValueOf(InventoryGroup.SHIELD, CalcType.PROTECTION)));
        table.add("").row();

        table.add(CalcType.DEFENSE.getTitle());
        table.add(String.valueOf(selectedHero.getCalcValueOf(InventoryGroup.SHIELD, CalcType.DEFENSE)));
        table.add("").row();

        table.add("Spell Battery");
        table.add("?");
        table.add("").row();
    }

}

File: core/src/nl/t64/game/rpg/screens/inventory/CalcsTable.kt
package nl.t64.game.rpg.screens.inventory

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import nl.t64.game.rpg.Utils.createTopBorder
import nl.t64.game.rpg.components.party.CalcAttributeId
import nl.t64.game.rpg.components.party.PersonalityItem
import nl.t64.game.rpg.components.party.inventory.InventoryGroup
import nl.t64.game.rpg.components.party.toCalcAttributeId
import nl.t64.game.rpg.constants.Constant
import nl.t64.game.rpg.screens.inventory.tooltip.PersonalityTooltip


private const val FIRST_COLUMN_WIDTH = 190f
private const val SECOND_COLUMN_WIDTH = 40f
private const val THIRD_COLUMN_WIDTH = 35f

internal class CalcsTable(tooltip: PersonalityTooltip) : BaseTable(tooltip) {

    init {
        table.columnDefaults(0).width(FIRST_COLUMN_WIDTH)
        table.columnDefaults(1).width(SECOND_COLUMN_WIDTH)
        table.columnDefaults(2).width(THIRD_COLUMN_WIDTH)

        container.add(table)
        container.background = createTopBorder()
        container.addListener(ListenerKeyVertical { updateIndex(it) })
    }

    private fun updateIndex(deltaIndex: Int) {
        selectedIndex += deltaIndex
        if (selectedIndex < 0) {
            selectedIndex = table.rows - 1
        } else {
            if (selectedIndex >= table.rows) {
                selectedIndex = 0
            }
        }
        hasJustUpdated = true
    }

    override fun fillRows() {
        table.add(Label(CalcAttributeId.WEIGHT.title, createLabelStyle()))
        table.add(selectedHero.getTotalCalcOf(CalcAttributeId.WEIGHT).toString())
        table.add("").row()

        table.add(Label(CalcAttributeId.MOVEPOINTS.title, createLabelStyle()))
        table.add(selectedHero.getCalculatedMovepoints().toString())
        addExtraToTable(selectedHero.getTotalCalcOf(CalcAttributeId.MOVEPOINTS))

        table.add(Label(CalcAttributeId.BASE_HIT.title, createLabelStyle()))
        table.add(selectedHero.getCalcValueOf(InventoryGroup.WEAPON, CalcAttributeId.BASE_HIT).toString() + "%")
        table.add("").row()

        table.add(Label("Total Hit", createLabelStyle()))
        table.add("?")
        table.add("").row()

        table.add(Label("Weapon " + CalcAttributeId.DAMAGE.title, createLabelStyle()))
        table.add(selectedHero.getCalcValueOf(InventoryGroup.WEAPON, CalcAttributeId.DAMAGE).toString())
        table.add("").row()

        table.add(Label("Total " + CalcAttributeId.DAMAGE.title, createLabelStyle()))
        table.add(selectedHero.getCalculatedTotalDamage().toString())
        table.add("").row()

        table.add(Label("Shield " + CalcAttributeId.DEFENSE.title, createLabelStyle()))
        table.add(selectedHero.getCalcValueOf(InventoryGroup.SHIELD, CalcAttributeId.DEFENSE).toString())
        table.add("").row()

        table.add(Label("Shield " + CalcAttributeId.PROTECTION.title, createLabelStyle()))
        table.add(selectedHero.getCalcValueOf(InventoryGroup.SHIELD, CalcAttributeId.PROTECTION).toString())
        table.add("").row()

        table.add(Label("Total " + CalcAttributeId.PROTECTION.title, createLabelStyle()))
        table.add(selectedHero.getTotalCalcOf(CalcAttributeId.PROTECTION).toString())
        addExtraToTable(selectedHero.getPossibleExtraProtection())

        table.add(Label(CalcAttributeId.SPELL_BATTERY.title, createLabelStyle()))
        table.add(selectedHero.getTotalCalcOf(CalcAttributeId.SPELL_BATTERY).toString())
        table.add("").row()

        if (table.hasKeyboardFocus()) {
            setSelected()
        }
    }

    private fun setSelected() {
        val child = table.getChild(selectedIndex * 3) as Label
        child.style.fontColor = Constant.DARK_RED
        if (hasJustUpdated) {
            hasJustUpdated = false
            tooltip.setPosition(FIRST_COLUMN_WIDTH / 1.5f) { getTooltipY() }
            tooltip.refresh(child, getPersonalityItemForDescriptionOnly(child))
        }
    }

    private fun getPersonalityItemForDescriptionOnly(child: Label): PersonalityItem {
        return object : PersonalityItem {
            override fun getDescription(totalLoremaster: Int): String {
                return child.text.toString().toCalcAttributeId()
                    ?.getDescription()
                    ?: "ToDo from CalcsTable"
            }
        }
    }

    private fun createLabelStyle(): LabelStyle {
        return LabelStyle(font, Color.BLACK)
    }

    private fun getTooltipY(): Float {
        val rowHeight = table.getRowHeight(0)
        return container.height - (rowHeight * selectedIndex) - (rowHeight * 0.5f)
    }

}
